core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -2953,6 +2979,166 @@ private void scheduleAsyncPersistenceInternal(LockedInodePath inodePath) throws
+    if (!mUfsSyncPathCache
+        .shouldSyncPath(path.getPath(), options.getCommonOptions().getSyncIntervalMs())) {
+      return false;
+    }
+
+    // True if ufs metadata should be loaded later.
+    boolean loadMetadata = false;
+
+    DeleteOptions syncDeleteOptions =
+        DeleteOptions.defaults().setRecursive(true).setAlluxioOnly(true).setUnchecked(true);
+
+    // The high-level process for the syncing is:
+    // 1. Find all Alluxio paths which are not consistent with the corresponding UFS path.
+    //    This means the UFS path does not exist, or is different from the Alluxio metadata.
+    // 2. Delete those Alluxio paths which are not consistent with Alluxio. After this step, all
+    //    the paths in Alluxio are consistent with UFS, and there may be additional UFS paths to
+    //    load.
+    // 3. Load metadata from UFS.
+
+    try (JournalContext journalContext = createJournalContext();
+         LockedInodePath inodePath = mInodeTree.lockInodePath(path, InodeTree.LockMode.WRITE);
+         FileSystemMasterAuditContext auditContext =
+             createAuditContext("syncMetadata", path, null, inodePath.getInodeOrNull())) {
+      try {
+        mPermissionChecker.checkPermission(Mode.Bits.READ, inodePath);
+      } catch (AccessControlException e) {
+        auditContext.setAllowed(false);
+        throw e;
+      }
+
+      if (!inodePath.fullPathExists()) {
+        // The requested path does not exist in Alluxio, so just load metadata.
+        loadMetadata = true;
+      } else {
+        // The requested path already exists in Alluxio.
+
+        MountTable.Resolution resolution = mMountTable.resolve(path);
+        AlluxioURI ufsUri = resolution.getUri();
+        UnderFileSystem ufs = resolution.getUfs();
+
+        UfsSyncUtils.SyncPlan syncPlan = UfsSyncUtils
+            .computeSyncPlan(inodePath.getInode(), ufs.getFingerprint(ufsUri.toString()));
+
+        if (syncPlan.toDelete() && inodePath.getParentInodeOrNull() != null) {
+          // Do not delete the root.
+          try {
+            deleteInternal(inodePath, true, System.currentTimeMillis(), syncDeleteOptions,
+                journalContext);
+          } catch (DirectoryNotEmptyException | IOException e) {
+            // Should not happen, since it is an unchecked delete.
+          }
+        }
+
+        if (syncPlan.toLoadMetadata()) {
+          loadMetadata = true;
+        }
+
+        if (syncPlan.toSyncChildren() && inodePath.getInode().isDirectory()) {
+          UfsStatus[] listStatus = null;
+          try {
+            listStatus = ufs.listStatus(ufsUri.toString());
+          } catch (IOException e) {
+            // ignore error
+            LOG.warn("Failed to list directory: {} error: {}", ufsUri, e.getMessage());
+          }
+
+          if (listStatus != null) {
+            // maps of children name to ufs or inode information
+            HashMap<String, String> ufsChildren = new HashMap<>();
```

Comment:
perhaps rename to `ufsFingerprints` or `ufsChildFingerprints`? 

`String ufsFingerprint = ufsChildren.get(inodeEntry.getKey())` looks suspicious

Reply:
Done